tutel/custom/custom_kernel.cpp

```diff
@@ -444,6 +460,54 @@ static void nccl_all_to_all_gather_async(
   g_cuda_events[0].record(get_nccl_stream());
 }
 
+static void all_to_all(torch::Tensor &output, torch::Tensor &input, const char *algo) {
+  CHECK_CUDA(output);
+  CHECK_CUDA(input);
+  auto recvbuff = (void*)output.data_ptr();
+  auto sendbuff = (void*)input.data_ptr();
+  cudaStream_t stream = get_nccl_stream().stream();
+
+  size_t length = input.nbytes();
+  CHECK_EQ(0, length % g_world_size);
+  size_t slice_size = length / g_world_size;
+
+  int nranks = g_world_size, ngpus = g_local_size;
+  CHECK_EQ(0, nranks % ngpus);
+  int nnodes = nranks / ngpus;
+  if (ngpus == 1 || nnodes == 1) goto linear;
+
+  if (algo && !strcmp(algo, "2D")) {
+    int node_rank = g_world_rank / ngpus, local_rank = g_local_rank;
+    // phase 0. per-gpu (ngpus) stride copy
+    CHECK_EQ(0, memStrideCopy(recvbuff, sendbuff, slice_size, ngpus, nnodes, stream));
+    // phase 1. intra-node alltoall
+    CHECK_EQ(0, ncclGroupStart());
+    for (int g = 0; g < ngpus; g++) {
+      CHECK_EQ(0, ncclSend(((char*)recvbuff) + g * nnodes * slice_size, nnodes * slice_size, ncclInt8, g + node_rank * ngpus, g_nccl_comm, stream));
+      CHECK_EQ(0, ncclRecv(((char*)sendbuff) + g * nnodes * slice_size, nnodes * slice_size, ncclInt8, g + node_rank * ngpus, g_nccl_comm, stream));
+    }
+    CHECK_EQ(0, ncclGroupEnd());
+    // phase 2. per-gpu (nnodes) stride copy
+    CHECK_EQ(0, memStrideCopy(recvbuff, sendbuff, slice_size, nnodes, ngpus, stream));
```

Comment:
We should keep input untouched, else we need to explicitly say in function name that this function destroys input.

Reply:
updated in pybind function description
I think we don't provide a coll interface, it's only used in tutel layer, reserving input will have extra copy/memory overhead